### ISSUE TITLE
Sending NICK and USER in single request

### ISSUE THREAD

**Issue by photinos** [hexchat/hexchat#1852]
In src/common/proto-irc.c in irc_login(..).  "NICK" & "USER" are being sent in single request line to server.  proto-irc.c lines 59-62.  This is causing login to hang on some IRC servers.  Needs to be sent a s two separate requests to server. 
i.e.
tcp_sendf (serv, "NICK %s\r\n",serv->nick );
tcp_sendf (serv, "USER %s %s %s :%s\r\n", user, user, serv->servername, realname);

**Comment by Arnavion** [comment]
They're being sent in two lines. There's a `\r\n` after the `NICK %s`.

If you meant to say that they should be sent in two separate calls to `tcp_sendf`, that makes no sense because it's all over the same socket. A server couldn't distinguish between the two approaches even if it wanted to.

**Comment by photinos** [comment]
Two separate calls to tcp_sendf  is what I am saying.  For some reason starting about 3-4+ months ago we started seeing this behavior on multiple different versions and types/vendors of IRC servers.  When connecting to some servers,  when the NICK and USER are sent within a  single tcp_sendf the sever gets and ACKs the NICK but never ACKS the USER and the client hangs trying to login.  We verified this by using Wireshark to watch the exchange.  Splitting out into two tcp_sendf corrected the issue.  Whether or not it make sense (and I agree with you and see the newline in the code) or not, this is an issue and simple fix that we implemented.  This does not effect all IRC servers, but a majority  of the common ones we connect to.

**Comment by Arnavion** [comment]
Who is "we" and what servers? The code has been there for years, so either you're connecting to some very esoteric servers or something's wrong with the network between you and the server.

**Comment by photinos** [comment]
Is there a way to continue this discussion off list?

**Comment by TingPing** [comment]
Surely its a symptom of a larger problem in the server, I don't see any point of the client working around it as its not even a behavioral difference its a bug.

**Comment by photinos** [comment]
Can we take this discussion off list?


Arnav Singh wrote:

**Comment by photinos** [comment]
Sir,
     Two separate calls to tcp_sendf  is what I am saying.  For some 
reason starting about 3-4+ months ago we started seeing this behavior on 
multiple different versions and types/vendors of IRC servers.  When 
connecting to some servers,  when the NICK and USER are sent within a  
single tcp_sendf the sever gets and ACKs the NICK but never ACKS the 
USER and the client hangs trying to login.  We verified this by using 
Wireshark to watch the exchange.  Splitting out into two tcp_sendf 
corrected the issue. Whether or not it make sense (and I agree with you 
and see the newline in the code) or not, this is an issue and simple fix 
that we implemented.  This does not effect all IRC servers, but a 
majority  of the common ones we connect to.

V/R,
     Bill


Arnav Singh wrote:
